### DISCUSSION TITLE
[Share] Copy and paste tags of selected item(s) 复制粘贴所选条目的标签

### DISCUSSION THREAD

**Discussion by redleafnew** [windingwind/zotero-actions-tags#194] [Action Scripts]
### Description

Copy tags from item and paste them to another item.

复制、粘贴所选条目的标签
大部分代码来源于Zutilo。
感谢@windingwind指导。

### Event

None

### Operation

Script

### Data

Action Data for **COPY** tags from item:

```javascript
// Copy tags. 
// Codes from Zutilo.
// 获取条目tags，并存入剪贴板
var tagsArray = [];
var items = Zotero.getActiveZoteroPane().getSelectedItems();
for (let item of items) {

    if (item.isRegularItem() && !(item instanceof Zotero.Collection)) {

        var tempTags = item.getTags();
        var arrayStr = '';
        for (var j = 0; j < tempTags.length; j++) {
            arrayStr = '\n' + tagsArray.join('\n') + '\n';

            let tag
            tag = tempTags[j].tag

            if (arrayStr.indexOf('\n' + tag + '\n') == -1) {
                tagsArray.push(tag);
            }
        }
    }

}
var clipboardText = tagsArray.join('\r\n');
Zotero.Utilities.Internal.copyTextToClipboard(clipboardText);
return "Tags copied."


```

Action Data for **PASTE** tags to item:

```javascript
// Paste tags
// Codes from Zutilo.
// 从剪贴板获取tags
var clipboardText = Zotero.Utilities.Internal.getClipboard("text/plain").trim()
if (!clipboardText) {
    return false;
}
// 拆分
var tagArray = clipboardText.split(/\r\n?|\n/)
tagArray = tagArray.map(function (val) { return val.trim() });
tagArray = tagArray.filter(Boolean)
// 添加tags
var items = Zotero.getActiveZoteroPane().getSelectedItems();
for (let item of items) {
    if (item.isRegularItem() && !(item instanceof Zotero.Collection)) {
        for (let tag of tagArray) {
            item.addTag(tag)
        }
        item.saveTx()
    }

}
return "Tags pasted."

```

### Anything else

Note that you'll need to create **TWO** actions, one for copy and one for paste.

有些代码可能不需要，可以自行删除。
需要分别建立两个动作。

**Comment by redleafnew** [comment]
https://getquicker.net/Sharedaction?code=16252683-cae2-4ced-492c-08db697c985a
这个Quicker动作也可以完成这些动作。

**Comment by Dcrizz** [comment]
Thanks very much for this; it works when I select the top-level/parent item, but when I try to copy the tags just for a note item (either below the parent item or elsewhere) it returns this (see the screenshot below), am I doing something wrong or do I need to modify the template?

![Screenshot 2024-01-10 at 18 13 05](https://github.com/windingwind/zotero-actions-tags/assets/121730622/23520b40-90b4-4158-a55d-7de57f1ed058)

These are the current settings I have:

![Screenshot 2024-01-10 at 18 17 38](https://github.com/windingwind/zotero-actions-tags/assets/121730622/b598e821-3476-4f01-9ed7-e5d7255ecdde)

Again, thank you  if you manage to take a look at this at all!!

**Comment by dschaehi** [comment]
Thanks for the action. I started getting this error recently (e.g. `7.0.0-beta.74+2e8073ab9`) when pasting from clipboard:
```
Script Error:
Cc['@mozilla.org/embedcomp/prompt-service;1'] is undefined
```
Do you have an idea how to fix this?

**Reply by richardlm** [reply]
I am getting the same error message.

**Comment by redleafnew** [comment]
updated

**Reply by dschaehi** [reply]
It works!

**Comment by mjthoraval** [comment]
It is not working for me to copy or paste tags on Standalone notes.

**Comment by redleafnew** [comment]
I think this script only works for regular item tag.

**Comment by dschaehi** [comment]
For the newest Zotero version I had to change `item.isCollection()` to `item instanceof Zotero.Collection`.

**Reply by redleafnew** [reply]
> For the newest Zotero version I had to change `item.isCollection()` to `item instanceof Zotero.Collection`.

Thank a lot. Updated.

**Comment by wuyulele** [comment]
thanks a lot, it used to work, but now it doens't work for 7.0.0-beta.95+694ccecf4. Could you please update the script?

**Comment by redleafnew** [comment]
For the newest Zotero version I had to change item.isCollection() to item instanceof Zotero.Collection.

**Comment by rodrigolourencofarinha** [comment]
Here is an updated script working with `7.0.0-beta.101+220bb4383 (64-bit)`

```
// Paste tags
// Codes from Zutilo.

var clipboardText = Zotero.Utilities.Internal.getClipboard("text/plain").trim()
if (!clipboardText) {
    return false;
}

var tagArray = clipboardText.split(/\r\n?|\n/)
tagArray = tagArray.map(function (val) { return val.trim() });
tagArray = tagArray.filter(Boolean)

var items = Zotero.getActiveZoteroPane().getSelectedItems();
for (let item of items) {
    if (item.isRegularItem() && !(item instanceof Zotero.Collection)) {
        for (let tag of tagArray) {
            item.addTag(tag)
        }
        item.saveTx()
    }

}
return "Tags pasted"
```

**Comment by yzotero** [comment]
Thanks a lot but the script does not allow to copy and paste tags from the notes in Zotero 7. Updated version would be very much appreciated! Thank you

**Comment by redleafnew** [comment]
有什么提示吗？

**Reply by yzotero** [reply]
Yes, would you know how to do copy and paste the tags from notes? Thanks in advance

**Reply by redleafnew** [reply]
当前代码只是针对变通条目的，如果对笔记的知，试试把两个代码中的
`item.isRegularItem() && `
删除试试

**Reply by yzotero** [reply]
Thanks, unfortunately I don't. Zutilo was great, it allows to copy and paste any tags from anywhere. If you find the script, please let us know. Thanks much in advance!

**Reply by redleafnew** [reply]
I think Zutilo for zoter 7 will release soon.

**Reply by redleafnew** [reply]
> Thanks, unfortunately I don't. Zutilo was great, it allows to copy and paste any tags from anywhere. If you find the script, please let us know. Thanks much in advance!

https://github.com/wshanks/Zutilo/issues/248#issuecomment-2281749673

**Reply by Dcrizz** [reply]
I managed to get this to work for copying and pasting tags for any item

### Copy Tags


```
// Copy tags.
// Codes from Zutilo.
// 获取条目tags，并存入剪贴板
var tagsArray = [];
var items = Zotero.getActiveZoteroPane().getSelectedItems();
for (let item of items) {
    // Check if the item is not a collection
    if (!item.isAttachment() && !item.isAnnotation()) {
        var tempTags = item.getTags();
        for (var j = 0; j < tempTags.length; j++) {
            let tag = tempTags[j].tag;
            if (!tagsArray.includes(tag)) {
                tagsArray.push(tag);
            }
        }
    }
}
var clipboardText = tagsArray.join('\r\n');
Zotero.Utilities.Internal.copyTextToClipboard(clipboardText);
return "Tags copied.";
```


### Paste Tags:

```
// Paste tags
// Codes from Zutilo.
// 从剪贴板获取tags
var clipboardText = Zotero.Utilities.Internal.getClipboard("text/plain").trim();
if (!clipboardText) {
    return false;
}
// 拆分
var tagArray = clipboardText.split(/\r\n?|\n/);
tagArray = tagArray.map(function (val) { return val.trim(); });
tagArray = tagArray.filter(Boolean);
// 添加tags
var items = Zotero.getActiveZoteroPane().getSelectedItems();
for (let item of items) {
    // Check if the item is not a collection
    if (!item.isAttachment() && !item.isAnnotation()) {
        for (let tag of tagArray) {
            item.addTag(tag);
        }
        item.saveTx();
    }
}
return "Tags pasted.";
```

Hope that helps

**Reply by yzotero** [reply]
You are my savior. Thank you very much!

**Reply by EureDudeheit** [reply]
Thank you, this works now perfect. 
@redleafnew maybe you could update your initial posting?

**Comment by redleafnew** [comment]
`!(item instanceof Zotero.Collection)` was used to replace `!item.isCollection()`, now it works.

**Comment by Trebuchosaure** [comment]
Hi, thank you very much for the scripts

I get an warning message when I created both actions

> This script is using 'ZoteroPane.getSelectedltems`, which is NOT recommended for getting items in scripts and can lead to unexpected behavior. Please use the 'item' or 'item' variables passed to the script instead.

**Reply by QuanzhiX** [reply]
Try my posted code

**Reply by Trebuchosaure** [reply]
Hello,
It works perfectly, thanks !

**Comment by laroospare** [comment]
Hello! This worked for me a few days ago; now when I try it returns script error Zotero.Utilities.Internal.getClipboard (...) is null.

Thanks for all your work on this.

**Comment by QuanzhiX** [comment]
```
// Paste tags
// Codes from Zutilo.
// 从剪贴板获取tags

// Ensure 'items' is defined (passed to the script)
if (typeof items === 'undefined' || !items.length) {
    Zotero.debug("No items selected.");
    return false;
}

var clipboardText = Zotero.Utilities.Internal.getClipboard("text/plain").trim();
if (!clipboardText) {
    Zotero.debug("Clipboard is empty.");
    return false;
}

// Split the clipboard text into an array of tags
var tagArray = clipboardText.split(/\r\n?|\n/);
tagArray = tagArray.map(function (val) { return val.trim(); });
tagArray = tagArray.filter(Boolean);

// Add tags to each selected item
for (let item of items) {
    if (item.isRegularItem() && !(item instanceof Zotero.Collection)) {
        for (let tag of tagArray) {
            item.addTag(tag);
        }
        item.saveTx();
    }
}

return "Tags pasted.";

```

**Reply by QuanzhiX** [reply]
Sort out the following:

> This script is using 'ZoteroPane.getSelectedltems`, which is NOT recommended for getting items in scripts and can lead to unexpected behavior. Please use the 'item' or 'item' variables passed to the script instead.

**Reply by amarois** [reply]
thanks; I've got the same message "This script is using 'ZoteroPane.getSelectedltems`, which is NOT recommended for getting items in scripts and can lead to unexpected behavior. Please use the 'item' or 'item' variables passed to the script instead." for the Copy script. Did you update this one to ?

**Reply by QuanzhiX** [reply]
Yes

**Reply by bfreed3** [reply]
Hi all, I'm just not getting this. Maybe I'm just not versed enough in coding this. I installed the Actions and Tags extension. But in addition to the codes, I also need to know all the other Actions and Tags options to choose for the Copy Tags and the Paste Tags,  most notably Operation, Event Menu Label options. . . So this above script works to paste, but what's the parallel script for copying that goes with this? Thanks,

**Reply by cpsyctc2** [reply]
I am trying to get my head around this but I am still getting the ``error message noted above when I try to create the copy tags action copying the latest copy tags code above.  I am new to the plugin and the coding here so I'm at a loss.

**Comment by adrianpip2000** [comment]
The following two scripts (copied and modified from previous comments) work with Zotero 7.0.11 + Actions & Tags build 2.1.0, and do not cause the warning message `This script is using 'ZoteroPane.getSelectedltems', which is NOT recommended (...)`.

### Copy tags
```JavaScript
// Copy tags script for Zotero Actions & Tags + Zotero 7
// Code from Zutilo (with modifications)
// 获取条目tags，并存入剪贴板

// Ensure 'items' is defined (passed to the script)
if (typeof items === 'undefined' || !items.length) {
    Zotero.debug("No items selected.");
    return false;
}

// Create empty array of tags
var tagArray = [];

// Copy tags from selected items into array
for (let item of items) {
    if (item.isRegularItem() && !(item instanceof Zotero.Collection)) {
        var tempTags = item.getTags();
        for (var j = 0; j < tempTags.length; j++) {
            let tag = tempTags[j].tag;
            if (!tagArray.includes(tag)) {
                tagArray.push(tag);
            }
        }
    }
}

// Copy tags to clipboard as text
var clipboardText = tagArray.join('\r\n');
Zotero.Utilities.Internal.copyTextToClipboard(clipboardText);

return "Tags copied.";
```

### Paste tags
```JavaScript
// Paste tags script for Zotero Actions & Tags + Zotero 7
// Code from Zutilo (with modifications)
// 从剪贴板获取tags

// Ensure 'items' is defined (passed to the script)
if (typeof items === 'undefined' || !items.length) {
    Zotero.debug("No items selected.");
    return false;
}

// Get tags from clipboard (but abort if empty)
var clipboardText = Zotero.Utilities.Internal.getClipboard("text/plain").trim();
if (!clipboardText) {
    Zotero.debug("Clipboard is empty.");
    return false;
}

// Split the clipboard text into an array of tags
var tagArray = clipboardText.split(/\r\n?|\n/);
tagArray = tagArray.map(function (val) { return val.trim(); });
tagArray = tagArray.filter(Boolean);

// Add tags to each selected item
for (let item of items) {
    if (item.isRegularItem() && !(item instanceof Zotero.Collection)) {
        for (let tag of tagArray) {
            item.addTag(tag);
        }
        item.saveTx();
    }
}

return "Tags pasted.";
```
